### PR TITLE
Fixed #30451 -- Added ASGI handler and coroutine-safety.

### PR DESCRIPTION
**Pull request by andrewgodwin**
This adds in an `AsgiHandler` handler class, `get_asgi_application`, and an `asgi.py` file into the default project template. It serves Django over an ASGI interface, but does not yet allow asynchronous views; `get_response` is where the context changes from asynchronous to synchronous.

It also swaps out Django's `threading`-based safety for `asgiref.local`, which provides hybrid thread- and async-task-based safety.

The goal would be to land this in `master`, and then start work on getting async through the middleware layers and down to views if they are declared as `async def`. That way we can stabilise this, make sure it's OK, and hopefully ship it in 3.0, and then see how far we can get async view handling as another PR so it's not all tied up in one big patch.

This will be squashed down to a single commit for merge.

### CODE REVIEW

**Comment by Bounder** [comment]
🎉

File: docs/spelling_wordlist [hunk 1]
```diff
@@ -789,3 +790,4 @@ xgettext
 XSS
 xxxxx
 Zope
+à
```
Comment (Ian-Foote) [inline docs/spelling_wordlist]: Is this file automatically ordered? These changes seem out of scope otherwise.

File: docs/howto/deployment/asgi/uvicorn.txt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+=========================
+
+When Uvicorn is installed, a ``uvicorn`` command is available which runs
+ASGI applications. Uvicorn needs to be called with the location of a
+module containing a ASGI application object, followed by what the application
+is called (separated by a colon).
+
+For a typical Django project, invoking Uvicorn would look like::
+
+    daphne myproject.asgi:application
```
Comment (ljodal) [inline docs/howto/deployment/asgi/uvicorn.txt]: Seems like this should say `uvicorn ...`?

Comment (andrewgodwin) [reply docs/howto/deployment/asgi/uvicorn.txt]: Yup!

File: docs/spelling_wordlist (hunk 1 above)
Comment (andrewgodwin) [reply docs/spelling_wordlist]: I automatically ordered it in my editor, but apparently that was a Bad Idea. I restored the old ordering instead so the diff is less polluted.

File: django/core/handlers/asgi.py [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+        if isinstance(response, FileResponse):
+            response.block_size = self.chunk_size
+        # Send the response
+        await self.send_response(response, send)
+
+    async def read_body(self, receive):
+        """
+        Reads a HTTP body from an ASGI connection.
+        """
+        body = b""
```
Comment (bluetech) [inline django/core/handlers/asgi.py]: Here you have a choice between `bytes` and `bytearray`. I ran this crude benchmark (Python 3.7, Linux, glibc):

```python
import time

CHUNK = b'\x00' * 1000
NUM_ITERS = 1000000

print('num_chunks,bytes_time,bytearray_time')
for num_chunks in range(1, 21):
    start = time.monotonic()
    for i in range(NUM_ITERS):
        body = b''
        for i in range(num_chunks):
            body += CHUNK
    finish = time.monotonic()
    bytes_time = time.monotonic() - start

    start = time.monotonic()
    for i in range(NUM_ITERS):
        body = bytearray()
        for i in range(num_chunks):
            body += CHUNK
        body = bytes(body)
    bytearray_time = time.monotonic() - start
    print(f'{num_chunks},{bytes_time},{bytearray_time}')
```

The result is:

bench

At least with the parameters I used, bytearray is slower for a small number of chunks, but looks linear, while bytes looks quadratic.

File: django/core/handlers/asgi.py [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+        # Ensure query string is encoded correctly
+        query_string = self.scope.get("query_string", "")
+        if isinstance(query_string, bytes):
+            query_string = query_string.decode("utf-8")
+        self.META = {
+            "REQUEST_METHOD": self.method,
+            "QUERY_STRING": query_string,
+            "SCRIPT_NAME": self.script_name,
+            "PATH_INFO": self.path_info,
+            # WSGI-epecting code will need these for a while
```
Comment (tomdyson) [inline django/core/handlers/asgi.py]: epecting -> expecting

File: django/core/handlers/asgi.py [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+            return self.request_class(scope, body), None
+        except UnicodeDecodeError:
+            logger.warning(
+                "Bad Request (UnicodeDecodeError)",
+                exc_info=sys.exc_info(),
+                extra={"status_code": 400},
+            )
+            return None, http.HttpResponseBadRequest()
+        except RequestTimeout:
+            # Parsing the rquest failed, so the response is a Request Timeout error
```
Comment (tomdyson) [inline django/core/handlers/asgi.py]: rquest -> request

File: docs/howto/deployment/asgi/daphne.txt [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+    pip install daphne
+
+
+Running Django in Daphne
+========================
+
+When Daphne is installed, a ``daphne`` command is available which starts
+the Gunicorn server process. At its simplest, Daphne just needs to be called
+with the location of a module containing a ASGI application object, followed
```
Comment (tomdyson) [inline docs/howto/deployment/asgi/daphne.txt]: a ASGI -> an ASGI

File: docs/howto/deployment/asgi/index.txt [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+Like WSGI, ASGI has you supply an ``application`` callable which
+the application server uses to communicate with your code. It's commonly
+provided as an object named ``application`` in a Python module accessible to
+the server.
+
+The :djadmin:`startproject` command creates a file
+:file:`<project_name>/asgi.py` that contains such an ``application`` callable.
+
+It's not used by the development server (``runserver``), but can be used by
+any ASGI server either in development or in  production.
```
Comment (tomdyson) [inline docs/howto/deployment/asgi/index.txt]: in  production -> in production (extra space)

File: docs/howto/deployment/asgi/uvicorn.txt [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+    pip install uvicorn
+
+
+Running Django in Uvicorn
+=========================
+
+When Uvicorn is installed, a ``uvicorn`` command is available which runs
+ASGI applications. Uvicorn needs to be called with the location of a
+module containing a ASGI application object, followed by what the application
```
Comment (tomdyson) [inline docs/howto/deployment/asgi/uvicorn.txt]: a ASGI -> an ASGI

File: django/core/handlers/asgi.py (hunk 3 above)
Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: Given this is likely going to be rewritten to use a temporary file in future, I will likely just leave it bytes for now.

**Comment by andrewgodwin** [comment]
buildbot, test this please

**Comment by andrewgodwin** [comment]
The MariaDB test appears to be consistently failing for all PRs today, so I am going to say this is otherwise ready for review.

File: django/core/handlers/asgi.py [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+        """
+        body = b""
+        while True:
+            message = await receive()
+            if message["type"] == "http.disconnect":
+                # Bye bye.
+                raise RequestAborted()
+            else:
+                # See if the message has body, and if it's the end, launch into
+                # handling (and a synchronous subthread)
```
Comment (jpic) [inline django/core/handlers/asgi.py]: I recon i'm a bit hesitating on this comment, does the following if-block concern "if it's the end" ? That's what i thought, but further down there is another " # Quit out if that's the end" comment so perhaps that one is more for the else-block than for the if-block.
But the next comment "Limit the maximum request data size" is above the if-block that it actually comments. 
Thanks for your support.

Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: You were very right - that comment leaked over from some old Channels code. I've revised it.

File: django/db/backends/postgresql/base.py [hunk 10]
```diff
@@ -224,12 +228,29 @@ def create_cursor(self, name=None):
         cursor.tzinfo_factory = utc_tzinfo_factory if settings.USE_TZ else None
         return cursor
 
+    @async_unsafe
     def chunked_cursor(self):
         self._named_cursor_idx += 1
+        # Get the current async task
```
Comment (claudep) [inline django/db/backends/postgresql/base.py]: async newbie question: can a current task exists in a method marked as `@async_unsafe`?

Comment (andrewgodwin) [reply django/db/backends/postgresql/base.py]: It cannot, but this is there for future-proofing should we decide to open up to async tasks (I basically just wanted to make sure as many uses of `threading` were covered as possible)

File: django/core/asgi.py [hunk 11]
```diff
@@ -0,0 +1,13 @@
+import django
+from django.core.handlers.asgi import AsgiHandler
+
+
+def get_asgi_application():
+    """
+    The public interface to Django's ASGI support. Return an ASGI 3 callable.
+
+    Avoids making django.core.handlers.ASGIHandler a public API, in case the
```
Comment (rsalmaso) [inline django/core/asgi.py]: django.core.handlers.AsgiHandler?

File: django/core/handlers/asgi.py [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+            self.META["REMOTE_ADDR"] = self.scope["client"][0]
+            self.META["REMOTE_HOST"] = self.META["REMOTE_ADDR"]
+            self.META["REMOTE_PORT"] = self.scope["client"][1]
+        if self.scope.get("server", None):
+            self.META["SERVER_NAME"] = self.scope["server"][0]
+            self.META["SERVER_PORT"] = str(self.scope["server"][1])
+        else:
+            self.META["SERVER_NAME"] = "unknown"
+            self.META["SERVER_PORT"] = "0"
+        # Handle old style-headers for a transition period
```
[the remark this answers is not in the document]
Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: Yeah, I think that can go!

File: docs/howto/deployment/asgi/daphne.txt [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+You can install Daphne with Pip::
+
+    pip install daphne
+
+
+Running Django in Daphne
+========================
+
+When Daphne is installed, a ``daphne`` command is available which starts
+the Gunicorn server process. At its simplest, Daphne just needs to be called
```
[the remark this answers is not in the document]
Comment (andrewgodwin) [reply docs/howto/deployment/asgi/daphne.txt]: Typo from copying the paragraph over!

File: django/db/backends/postgresql/base.py (hunk 10 above)
Comment (claudep) [reply django/db/backends/postgresql/base.py]: Ah thanks. But then I'd suggest to insert a (temporary) note explaining that to avoid a similar question later.

**Comment by pope1ni** [comment]
So I'm fairly ignorant of async stuff at moment, but I see two things crop up a lot here:

- `django.utils.asyncio.async_unsafe()`
- `asgiref.local.Local()` in place of `threading.local()`

I presume that third-party apps I use may need to make use of these to support running under ASGI?

In the case of the latter, is `asgiref.local.Local()` expected to replace all uses of `threading.local()`?

Given these questions, will there be some documentation on implementing minimal support for ASGI for apps?

File: django/db/backends/postgresql/base.py (hunk 10 above)
Comment (andrewgodwin) [reply django/db/backends/postgresql/base.py]: Will do.

**Comment by andrewgodwin** [comment]
@pope1ni If you want to be fully async-safe, then yes, it's required that you account for coroutines in a third-party app. `async_unsafe` is a general protection against this occurring if you have code you know is unsafe for sure, while `Local` is a helpful implementation if you want threadlocal-like behaviour with async coroutines.

Neither is required for writing ASGI apps at all, but we will need a guide that is "adapting Django apps to work under async" which mentions both of these, as well as a slew of other factors like "don't use normal, sync `requests` in async code" and "don't use normal sockets". I did not intend to include that in this patch, since it's far more extensive than this basic safety. I'll add a warning to the ASGI deployment docs that _are_ in here, though.

**Comment by pope1ni** [comment]
@andrewgodwin Thanks for the explanation. Looking forward to seeing how this develops.

File: django/contrib/staticfiles/handlers.py [hunk 14]
```diff
@@ -64,3 +72,35 @@ def __call__(self, environ, start_response):
         if not self._should_handle(get_path_info(environ)):
```
Comment (apollo13) [inline django/contrib/staticfiles/handlers.py]: This already checks for `_should_handle`. Why is this duplication also in `get_response`? Because if you eliminate the duplication the wsgi and asgi `get_response` are exactly the same and can be moved up into the common class, no?

File: django/core/handlers/asgi.py [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+
+    def __init__(self, scope, body):
+        self.scope = scope
+        self._content_length = 0
+        self._post_parse_error = False
+        self._read_started = False
+        self.resolver_match = None
+        self.script_name = self.scope.get("root_path", "")
+        if self.script_name and scope["path"].startswith(self.script_name):
+            # TODO: Better is-prefix checking, slash handling?
```
Comment (apollo13) [inline django/core/handlers/asgi.py]: More importantly I think our wsgi handler currently has the oddity to always normalize to at least /, you will never get an empty path_info iirc

File: django/core/handlers/asgi.py [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+        self.META = {
+            "REQUEST_METHOD": self.method,
+            "QUERY_STRING": query_string,
+            "SCRIPT_NAME": self.script_name,
+            "PATH_INFO": self.path_info,
+            # WSGI-expecting code will need these for a while
+            "wsgi.multithread": True,
+            "wsgi.multiprocess": True,
+        }
+        if self.scope.get("client", None):
```
Comment (apollo13) [inline django/core/handlers/asgi.py]: `scope` is a dict, isn't the `None` already the default anyways? Same below.

File: django/core/handlers/asgi.py [hunk 17]
```diff
@@ -0,0 +1,342 @@
...
+            if message["type"] == "http.disconnect":
+                # Early client disconnect. Bye bye.
+                raise RequestAborted()
+            else:
+                # If there is a body chunk in the message, add it to what we have.
+                if "body" in message:
+                    body += message["body"]
+                # Limit the maximum request data size that will be handled in-memory.
+                # TODO: Stream the body to temp disk instead?
+                # (we can't provide a file-like with direct reading as we would not be async)
```
Comment (apollo13) [inline django/core/handlers/asgi.py]: This probably cannot use the upload handlers for similar reasons?

File: django/contrib/staticfiles/handlers.py (hunk 14 above)
Comment (andrewgodwin) [reply django/contrib/staticfiles/handlers.py]: I agree, but considering the code was already like this, I did not want to adjust it without understanding why. Any ideas?

File: django/core/handlers/asgi.py (hunk 15 above)
Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: I believe this at least handles that - do you see a behaviour here that needs changing?

File: django/core/handlers/asgi.py (hunk 16 above)
Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: I often just put `None` in my `gets` just to make it clearer to everyone reading the code. I'll remove it.

File: django/core/handlers/asgi.py (hunk 17 above)
Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: I haven't looked into that yet, but probably not. File input in general has been a pain in Channels and I expect the same here.

**Comment by andrewgodwin** [comment]
I have added a Trac ticket to match this here: https://code.djangoproject.com/ticket/30451

File: django/contrib/staticfiles/handlers.py (hunk 14 above)
Comment (carltongibson) [reply django/contrib/staticfiles/handlers.py]: I think it can be removed: #11336

File: django/core/handlers/asgi.py [hunk 18]
```diff
@@ -0,0 +1,342 @@
...
+        self._stream = BytesIO(self._body)
+        # Other bits
+        self.resolver_match = None
+
+    @cached_property
+    def GET(self):
+        return http.QueryDict(self.scope.get("query_string", ""))
+
+    def _get_scheme(self):
+        return self.scope.get("scheme", "http")
```
[the remark this answers is not in the document]
Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: :+1:

File: django/core/handlers/asgi.py [hunk 19]
```diff
@@ -0,0 +1,342 @@
...
+            )
+        else:
+            self.path = scope["path"]
+
+        # HTTP basics
+        self.method = self.scope["method"].upper()
+        # Ensure query string is encoded correctly
+        query_string = self.scope.get("query_string", "")
+        if isinstance(query_string, bytes):
+            query_string = query_string.decode("utf-8")
```
[the remark this answers is not in the document]
Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: This is done so that `request.META['QUERY_STRING']` is always a consistent encoding - WSGI doesn't say that it's one way or the other, and in my experience people operate on it like a unicode string (especially as it's often used with `path`). Happy to change it to always be `bytes`, but I'd like to keep the consistent behaviour.

File: django/core/handlers/asgi.py [hunk 20]
```diff
@@ -0,0 +1,342 @@
...
+        self._body = body
+        assert isinstance(self._body, bytes), "Body is not bytes"
+        # Add a stream-a-like for the body
+        self._stream = BytesIO(self._body)
+        # Other bits
+        self.resolver_match = None
+
+    @cached_property
+    def GET(self):
+        return http.QueryDict(self.scope.get("query_string", ""))
```
[the remark this answers is not in the document]
Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: Right, this is a direct copy of what's in `WSGIHandler`, no mutable there. I'll fix the second argument to `get` though.

File: django/core/handlers/asgi.py [hunk 21]
```diff
@@ -0,0 +1,342 @@
...
+
+class AsgiHandler(base.BaseHandler):
+    """
+    Handler for ASGI requests.
+    """
+
+    request_class = AsgiRequest
+
+    # Size to chunk response bodies into for multiple response messages
+    chunk_size = 512 * 1024
```
[the remark this answers is not in the document]
Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: It's historical, from when the ASGI spec said you had to be under a megabyte. Will adjust to `2 ^ 16`.

File: django/db/backends/postgresql/base.py [hunk 22]
```diff
@@ -227,12 +231,32 @@ def create_cursor(self, name=None):
...
+        # Note that right now this is behind @async_unsafe, so this is
+        # unreachable, but in future we'll start loosening this restriction.
+        # For now, it's here so that every use of "threading" is also async-compatible.
+        try:
+            if hasattr(asyncio, "current_task"):
+                # Python 3.7 and up
+                current_task = asyncio.current_task()
+            else:
+                # Python 3.6
+                current_task = asyncio.Task.current_task()
```
[the remark this answers is not in the document]
Comment (andrewgodwin) [reply django/db/backends/postgresql/base.py]: Really? I've always preferred feature-testing to version-testing, especially as this would be the first time we need `PY37` in this file.

File: django/db/backends/postgresql/base.py [hunk 23]
```diff
@@ -227,12 +231,32 @@ def create_cursor(self, name=None):
...
+        except RuntimeError:
+            current_task = None
+        if current_task:
+            task_ident = str(id(current_task))
+        else:
+            task_ident = "sync"
```
[the remark this answers is not in the document]
Comment (andrewgodwin) [reply django/db/backends/postgresql/base.py]: Unfortunately I believe `asyncio.current_task` can return `None` _as well as_ raising a `RuntimeError`.

File: django/db/backends/postgresql/base.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (felixxm) [reply django/db/backends/postgresql/base.py]: Agreed, lets leave it. I thought that maybe it will be easier to remove Python 3.6-workarounds in the future.

File: django/db/backends/postgresql/base.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (felixxm) [reply django/db/backends/postgresql/base.py]: I was afraid that this can be the case :disappointed:, so maybe ternary :wink: 
```python
task_ident = str(id(current_task)) if current_task else 'sync'
```

File: django/db/backends/postgresql/base.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (andrewgodwin) [reply django/db/backends/postgresql/base.py]: Good point. Hopefully the comment with `3.6` in it will stand out enough.

File: django/db/backends/postgresql/base.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (andrewgodwin) [reply django/db/backends/postgresql/base.py]: Oh, interesting, I dislike ternary clauses (I think full `if` statements are easier to read). I did add a comment to make it clearer - can I convince you to leave it as an `if`? :)

[the remark this answers is not in the document]
Comment (felixxm) [reply django/db/backends/postgresql/base.py]: Sure, we can leave it :+1:

**Comment by andrewgodwin** [comment]
@carltongibson The `get_response` thing looks fine, happily pulled in. I'll re-squash this all down once the review changes are done.

The `qualname` thing is strange, I've not seen that here or in the Jenkins tests. I'll go looking for what it is, but without a Mac to replicate on, I may be a bit stuck.

@felixxm I have changed the quotes to single-quotes where it matches (e.g. the signals file was double-quotes already). Also reasonably sure I got all the comments under 79 but let me know if you spot any more.

File: django/core/handlers/asgi.py [hunk 24]
```diff
@@ -0,0 +1,332 @@
...
+        self._body = body
+        assert isinstance(self._body, bytes), 'Body is not bytes'
+        # Add a stream-a-like for the body
+        self._stream = BytesIO(self._body)
+        # Other bits
+        self.resolver_match = None
+
+    @cached_property
+    def GET(self):
+        return http.QueryDict(self.scope.get('query_string'))
```
Comment (felixxm) [inline django/core/handlers/asgi.py]: We process `query_string` in `__init__()`, so can we use instead `http.QueryDict(self.META['QUERY_STRING'])`?

Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: :+1:

**Comment by andrewgodwin** [comment]
@felixxm Yeah I was torn between matching `HttpRequest` and `WSGIHandler`. Don't mind either way.

I'm also going to see if I can add in proper request body handling to this before we land it now as that's maybe the only thing that has been bugging me.

**Comment by andrewgodwin** [comment]
There, I've updated it so it'll correctly handle large POST bodies and buffer them to disk once they go over `FILE_UPLOAD_MAX_MEMORY_SIZE`, which is the best we can do until maybe we prototype async-ingestion-hidden-behind-a-threaded-file-object over in Channels and see how stable it is in everyday usage. In the meantime, this pattern is well-known and what Django does currently anyway.

I did notice we have literally no tests for pushing up large uploaded files, mostly as we have very few tests that go through the actual Handlers rather than through the test Client (and its fake handler). I ran a full set of manual tests on this locally; I'm trying to figure out if there's a way to sensibly test this in the test suite without actually spinning up a server but it seems hard for both the WSGI and the ASGI case.

File: django/core/handlers/asgi.py [hunk 25]
```diff
@@ -0,0 +1,325 @@
...
+            self.META['SERVER_PORT'] = '0'
+        # Headers go into META
+        for name, value in self.scope.get('headers', []):
+            name = name.decode('latin1')
+            if name == 'content-length':
+                corrected_name = 'CONTENT_LENGTH'
+            elif name == 'content-type':
+                corrected_name = 'CONTENT_TYPE'
+            else:
+                corrected_name = 'HTTP_%s' % name.upper().replace('-', '_')
```
Comment (felixxm) [inline django/core/handlers/asgi.py]: Maybe:
```python
for name, value in self.scope.get('headers', []):
    corrected_name = name.decode('latin1').upper().replace('-', '_')
    if corrected_name not in ('CONTENT_LENGTH', 'CONTENT_TYPE')
        corrected_name = 'HTTP_%s' corrected_name
```

File: django/core/handlers/asgi.py [hunk 26]
```diff
@@ -0,0 +1,325 @@
...
+            )
+            if 'charset' in self.content_params:
+                try:
+                    codecs.lookup(self.content_params['charset'])
+                except LookupError:
+                    pass
+                else:
+                    self.encoding = self.content_params['charset']
+        else:
+            self.content_type, self.content_params = '', {}
```
Comment (felixxm) [inline django/core/handlers/asgi.py]: We have exactly the same (similar) code in `WSGIRequest.__init__()`. Maybe we can add an internal hook to the `HttpRequest`? (in a separate commit) e.g.
```python
def _parse_content_type_params(self, meta):
    self.content_type, self.content_params = cgi.parse_header(meta.get('CONTENT_TYPE', ''))
    if 'charset' in self.content_params:
        try:
            codecs.lookup(self.content_params['charset'])
        except LookupError:
            pass
        else:
            self.encoding = self.content_params['charset']
```
after that we could just call:
```python
self._parse_content_type_params(self.META)
```

Let me know if it works for you, I can handle this.

File: django/core/handlers/asgi.py [hunk 27]
```diff
@@ -0,0 +1,325 @@
...
+        # Pull out content length info
+        if self.META.get('CONTENT_LENGTH', None):
+            try:
+                self._content_length = int(self.META['CONTENT_LENGTH'])
```
Comment (felixxm) [inline django/core/handlers/asgi.py]: It seems that we doesn't use `self._content_length`. Do we need it?

File: django/core/handlers/asgi.py [hunk 28]
```diff
@@ -0,0 +1,325 @@
...
+from io import BytesIO
+
+from asgiref.sync import sync_to_async
+
+from django import http
+from django.conf import settings
+from django.core import signals
+from django.core.exceptions import RequestAborted, RequestDataTooBig
+from django.core.handlers import base
+from django.http import FileResponse, HttpResponse, HttpResponseServerError
```
Comment (felixxm) [inline django/core/handlers/asgi.py]: Can we import also `HttpRequest`, `HttpResponseBadRequest`, `parse_cookie`, and `QueryDict` and remove `from django import http`?

File: django/core/handlers/asgi.py [hunk 29]
```diff
@@ -0,0 +1,325 @@
...
+            body_file = await self.read_body(receive)
+        except RequestAborted:
+            return
+        # Request is complete and can be served.
+        set_script_prefix(self.get_script_prefix(scope))
+        await sync_to_async(signals.request_started.send)(sender=self.__class__, scope=scope)
+        # Get the request and check for basic issues
+        request, error_response = self.create_request(scope, body_file)
+        if request is None:
+            await self.send_response(error_response, send)
```
Comment (felixxm) [inline django/core/handlers/asgi.py]: I'm not sure why, but I would use `if error_response` instead of `if request is None`.

File: docs/releases/3.0.txt [hunk 30]
```diff
@@ -38,6 +38,29 @@ compatible with Django 3.0.
...
+you want to learn more.
+
+Note that as a side-effect of this change, Django is now aware of asynchronous
+event loops and will block you calling code marked as "async unsafe" - such as
+ORM operations - from an asynchronous context. If you were using Django from
+async code before, this may trigger if you were doing it incorrectly; if you
+see a ``SynchronousOnlyOperation`` error, then closely examine your code and
+move any database operations to be in a synchronous child thread.
+
+
```
Comment (felixxm) [inline docs/releases/3.0.txt]: Please chop one blank line.

File: django/core/handlers/asgi.py (hunk 25 above)
Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: I would say that the version I have is easier to read, though?

File: django/core/handlers/asgi.py (hunk 26 above)
Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: Yeah, I think we can clean this up later on. It's definitely in a couple of places.

File: django/core/handlers/asgi.py (hunk 27 above)
Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: Nope, it was used for a now-removed upload file size limit (that's handled elsewhere, better, now).

File: django/core/handlers/asgi.py (hunk 28 above)
Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: :+1:

File: django/core/handlers/asgi.py (hunk 29 above)
Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: I would prefer this, as, in my mind, "request is None" is the contract that says it's bad.

**Comment by andrewgodwin** [comment]
I have re-squashed after the latest round of tweaks requested by @felixxm - I feel like we're getting ready to land, but I'm going to wait until I get an explicit "yes" from both of the Fellows :)

File: django/core/handlers/asgi.py (hunk 26 above)
Comment (felixxm) [reply django/core/handlers/asgi.py]: I added this hook in #11471.

File: django/core/handlers/asgi.py (hunk 25 above)
Comment (felixxm) [reply django/core/handlers/asgi.py]: FYI: we have also [`HttpHeaders`](https://github.com/django/django/blob/fc2536fe66c519b306f673672b795d16f87ed57d/django/http/request.py#L359-L374) class that have reserve logic, I'm not sure if we can reuse it somehow :thinking:

File: django/core/handlers/asgi.py [hunk 31]
```diff
@@ -0,0 +1,320 @@
...
+    @cached_property
+    def GET(self):
+        return QueryDict(self.META['QUERY_STRING'])
+
+    def _get_scheme(self):
+        return self.scope.get('scheme') or super()._get_scheme()
+
+    def _get_post(self):
+        if not hasattr(self, '_post'):
+            self._read_started = False
```
[the remark this answers is not in the document]
Comment (andrewgodwin) [reply django/core/handlers/asgi.py]: Yup, I've removed it. This now leaves the `_post` functions the same as in WSGIHandler, but I don't want to move them into the base one as they weren't there already for a reason, I presume.

[the remark this answers is not in the document]
Comment (carltongibson) [reply django/core/handlers/asgi.py]: Yes, I agree. Thanks.

**Comment by carltongibson** [comment]
OK, this looks good to me. It's the start of a long road. 🙂

I kind of think that maybe we can tweak doc-here or doc-there but that it's pretty clear for the stage we're at. I think we should merge as-is and come back as the rest falls into place, and people start using it, particularly towards the end of the year. 

Outstanding question: do we want a merge #11471, and abstract that bit. (I don't really mind either way. DRY, yes, but sometimes these things are just better inlined. @felixxm you can decide. 🙂 And I assume you'll want to give it one last look over for remaining cosmetic changes?

**Comment by andrewgodwin** [comment]
I merge #11471 as @felixmm and I both agreed it should be spun out like that (and it's not like Request doesn't do that a lot already). Will rework this so it matches.

**Comment by felixxm** [comment]
@andrewgodwin I pushed some minor edits to comments and tests, added `asgiref` to the list of required packages for running all tests, and added two extra tests.

I'm going to check docs tomorrow and maybe add few more tests (if you prefer we can push them in a separate commit).

I have also one question about `ASGIStaticFilesHandler`, because currently it is unused. Can you give me some advice about it? How can I use it? :thinking:

**Comment by felixxm** [comment]
@andrewgodwin  Can you check `asgi.tests.ASGITest.test_file_response` failure on Windows?

**Comment by andrewgodwin** [comment]
@felixxm Do we need to add `asgiref` to the guide for running tests since the patch also introduces it as a Django dependency in `setup.py`?

Also, I'm looking into the test failure on Windows, it's an apparent lack of a system magic-numbers database. Will fix so it accepts both as valid mimetypes.

**Comment by andrewgodwin** [comment]
Also, ASGIStaticFilesHandler would in theory be used either in an `asgi.py` that was used to serve it in production, or an ASGI-enabled `runserver` (where the current static files handler is used).

**Comment by carltongibson** [comment]
> ...`ASGIStaticFilesHandler`...

Good yes. 🙂

So you'd do something like:

```
application = ASGIStaticFilesHandler(get_asgi_application())`
```

**Comment by felixxm** [comment]
> Do we need to add asgiref to the guide for running tests since the patch also introduces it as a Django dependency in setup.py?

Yes, we have all required packages there, also packages included in the `setup.py` like `sqlparse` or `pytz`.

**Comment by felixxm** [comment]
I squashed commits, pushed minor edits to docs and added more tests.

File: django/core/handlers/asgi.py [hunk 32]
```diff
@@ -0,0 +1,299 @@
...
+                corrected_name = 'CONTENT_LENGTH'
+            elif name == 'content-type':
+                corrected_name = 'CONTENT_TYPE'
+            else:
+                corrected_name = 'HTTP_%s' % name.upper().replace('-', '_')
+            # HTTP/2 say only ASCII chars are allowed in headers, but decode
+            # latin1 just in case.
+            value = value.decode('latin1')
+            if corrected_name in self.META:
+                value = self.META[corrected_name] + ',' + value
```
Comment (felixxm) [inline django/core/handlers/asgi.py]: @andrewgodwin I think these two lines can be removed because `corrected_name` cannot be in `META` at this point. What do you think?

Comment (carltongibson) [reply django/core/handlers/asgi.py]: Duplicate headers?

Comment (felixxm) [reply django/core/handlers/asgi.py]: Ahhh, yes, thanks!

**Comment by felixxm** [comment]
buildbot, test on selenium.

**Comment by felixxm** [comment]
buildbot, test on selenium.

**Comment by felixxm** [comment]
Many thanks Andrew :rocket:

**Comment by PetterS** [comment]
Whohoo! Great job!
